Config Tweak

nicegamer7 opened the pull request:
This tweaks how comments are added to the config files. It also fixes an issue with extra newlines being added in places they shouldn't. See below.

Before:
```
# Multiverse-Inventories Settings


# ===[ Multiverse Inventories Config ]===
settings:

  # This is the locale you wish to use.
  locale: en

  # If this is true it will generate world groups for you based on MV worlds.
  first_run: false

  # If this is set to true, it will enable bypass permissions (Check the wiki for more info.)
  use_bypass: false

  # If set to true, any world not listed in a group will automatically use the settings for the default group!
  default_ungrouped_worlds: false

  # The default and suggested setting for this is FALSE.
  # False means Multiverse-Inventories will not attempt to load or save any player data when they log in and out.
  # That means that MINECRAFT will handle that exact thing JUST LIKE IT DOES NORMALLY.
  # Changing this to TRUE will have Multiverse-Inventories save player data when they log out and load it when they log in.
  # The biggest potential drawback here is that if your server crashes, player stats/inventories may be lost/rolled back!
  save_load_on_log_in_out: false

  # If this is set to true, players will have different inventories/stats for each game mode.
  # Please note that old data migrated to the version that has this feature will have their data copied for both game modes.
  use_game_mode_profiles: false

shares:

  # When set to true, optional shares WILL be utilized in cases where a group does not cover their uses for a world.
  # An example of this in action would be an ungrouped world using last_location.  When this is true, players will return to their last location in that world.
  # When set to false, optional shares WILL NOt be utilized in these cases, effectively disabling it for ungrouped worlds.
  optionals_for_ungrouped_worlds: true

  # You must specify optional shares you wish to use here or they will be ignored.
  # The only built in optional share is "economy"
  use_optionals: []

```
```
# Multiverse-Inventories Groups


# To ADD, DELETE, and EDIT groups use the command /mvinv group.
# No support will be given for those who manually edit these groups.
groups:

  default:

    worlds:

    - world

    - world_the_end

    - world_nether

    shares:

    - ender_chest

    - inventory_contents

    - armor_contents

    - off_hand

    - hit_points

    - remaining_air

    - maximum_air

    - fall_distance

    - fire_ticks

    - xp

    - lvl

    - total_xp

    - food_level

    - exhaustion

    - saturation

    - bed_spawn

    - last_location

    - potion_effects

```

After:
```
# Multiverse-Inventories Settings
settings:
  # This is the locale you wish to use.
  locale: en

  # If this is true it will generate world groups for you based on MV worlds.
  first_run: false

  # If this is set to true, it will enable bypass permissions (Check the wiki for more info.)
  use_bypass: false

  # If set to true, any world not listed in a group will automatically use the settings for the default group!
  default_ungrouped_worlds: false

  # The default and suggested setting for this is FALSE.
  # False means Multiverse-Inventories will not attempt to load or save any player data when they log in and out.
  # That means that MINECRAFT will handle that exact thing JUST LIKE IT DOES NORMALLY.
  # Changing this to TRUE will have Multiverse-Inventories save player data when they log out and load it when they log in.
  # The biggest potential drawback here is that if your server crashes, player stats/inventories may be lost/rolled back!
  save_load_on_log_in_out: false

  # If this is set to true, players will have different inventories/stats for each game mode.
  # Please note that old data migrated to the version that has this feature will have their data copied for both game modes.
  use_game_mode_profiles: false

shares:
  # When set to true, optional shares WILL be utilized in cases where a group does not cover their uses for a world.
  # An example of this in action would be an ungrouped world using last_location.  When this is true, players will return to their last location in that world.
  # When set to false, optional shares WILL NOt be utilized in these cases, effectively disabling it for ungrouped worlds.
  optionals_for_ungrouped_worlds: true

  # You must specify optional shares you wish to use here or they will be ignored.
  # The only built in optional shares are "economy" and "last_location".
  use_optionals: []

```
```
# Multiverse-Inventories Groups
# To ADD, DELETE, and EDIT groups use the command /mvinv group.
# No support will be given for those who manually edit these groups.
groups:
  default:
    worlds:
    - world
    - world_the_end
    - world_nether
    shares:
    - hit_points
    - food_level
    - saturation
    - exhaustion
    - xp
    - total_xp
    - lvl
    - inventory_contents
    - armor_contents
    - bed_spawn
    - maximum_air
    - remaining_air
    - fall_distance
    - fire_ticks
    - potion_effects
    - last_location
    - ender_chest
    - off_hand

```

nicegamer7 commented:
Sorry, I forgot to note that this change does not cause any ill effects when updating from an older config. It just removes any extra newline characters if there are any.

dumptruckman commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -66,44 +69,36 @@ public boolean save() {
+            if (stringConfig.contains("\r\n")) lineEnding = "\r\n";
+            else lineEnding = "\n";
+
+            // convert stringConfig to array, ignoring the header
+            String[] arrayConfig = stringConfig.substring(stringConfig.indexOf(lineEnding) + lineEnding.length()).split(lineEnding);
+
+            // begin building the new config, starting with the header
             StringBuilder newContents = new StringBuilder();
-            String initialContents = config.options().header();
-            if (initialContents == null) {
-                initialContents = "";
-            }
-            newContents.append(initialContents)
-                    .append(System.getProperty("line.separator"))
-                    .append(System.getProperty("line.separator"));
+            newContents.append("# ").append(config.options().header()).append(lineEnding);
+
```
What if there are line breaks in the header?

nicegamer7 replied:
I'm unsure, I'll have to test it out and make sure it's handled correctly. Good catch!

nicegamer7 replied:
No, it won't be handled correctly, I'll have to make a small change.

nicegamer7 replied:
I fixed it, but now I'm experiencing some strange behavior which I can't understand. I changed the line endings to be LF to match [YamlConfiguration](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/configuration/file/YamlConfiguration.java#119) so that we won't end up making a file with mixed line endings. But if you enter a header with more than one line, the line endings between the lines will be CRLF? I can't figure out where the CR is coming from because it's not YamlConfiguration (the source for the header building method is linked above), and it's not Multiverse-Inventories since I replaced all `System.getProperty("line.separator")` with `\n`... so I'm not sure.

nicegamer7 replied:
Just figured it out. It has to do with feildmaster's Configuration-Library. Gonna see what can be done.

nicegamer7 replied:
Fixed it by removing Configuration-Library. MV Inventories is a little leaner now 😃.

dumptruckman commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -213,63 +190,52 @@ public boolean save() {
-            if (commentstring.length() > 0) {
-                commentstring.append("\r\n");
+            if (commentString.length() > 0) {
+                commentString.append(System.getProperty("line.separator"));
```
Isn't this the actual fix?

nicegamer7 replied:
No, I forget what the actual fix is though. I'll get back to you on this.

nicegamer7 replied:
The actual issue was the old line 73. The square brackets in the regex caused the resulting array to look something like this:
```
["Line 1...", "", "Line 2...", "", "Line 3", "", etc...
```
Removing the square brackets fixes the issue, then I added Lines 172 and 173 to keep gaps between config options with comments.

dumptruckman commented:
Very brave of you to modify CommentedYamlConfig. I've always considered it the most hard to refactor code I've ever written.

nicegamer7 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -66,44 +69,36 @@ public boolean save() {
+            // convert config file to String
+            String stringConfig = this.convertFileToString(file);
+
+            // detect which kind of line endings it uses
+            String lineEnding;
+            if (stringConfig.contains("\r\n")) lineEnding = "\r\n";
+            else lineEnding = "\n";
```
I added this to make it so that if a user moves their config files from a system using CRLF to a system using LF (or vice versa), they'd be converted. But now I'm wondering if the server does this automatically. What do you think?

SlimeDog replied:
> But now I'm wondering if the server does this automatically. What do you think?

If by "converted" you mean "the server reads them with error" = true. If you mean "files are converted to the type of the system" = false. At least in the DOS->Linux direction (ie., on Linux servers), files are not converted automatically. I convert every DOS-produced file with `dos2unix`. But I do that only for my own convenience and sanity; conversion is not required for the files to be read correctly.

nicegamer7 replied:
I see... if it's not done automatically, maybe we shouldn't be doing either.

nicegamer7 replied:
Looking at the source code of the [YamlConfiguration](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/browse/src/main/java/org/bukkit/configuration/file/YamlConfiguration.java#119) class, it seems Spigot has chosen to make all config files use the LF line ending, so we will do the the same.

nicegamer7 commented:
It wasn't too bad 😊.

nicegamer7 commented:
I just rebased this PR onto `main` and added two more commits. The first removes Configuration-Library, which was causing configs with multi-line headers to have mixed line-endings, and the second makes it so that `first_run` is always set to `false` after the first run of the plugin.

With that, this PR is ready for review!

I noticed one final issue, but it's an issue that was already present. If a configuration file cannot be read because it contains invalid YAML, then it is overwritten. I don't think it should be doing that... but I also don't want to fix it in this PR because the diff is beginning to get rather large. I'll open a follow-up PR later on.

benwoo1110 commented:
For the yaml being overwritten issue, I suggest copying the broken file to like a file named `<time>-config.yml.broken`.

Looking at the new config, do you want to retain the somewhat cooler `# ===[ Multiverse Inventories Config ]===` header instead of `# Multiverse-Inventories Settings`>

Also, do you think we should consider sharing this Comments in Config feature and use it in mv-portals and mv-core's configuration file? @nicegamer7 @dumptruckman what do you all think?

Anyways looks good to be me but I think this will need @dumptruckman approval before merging.

benwoo1110 commented:
Oh btw, do you think putting the wiki link to the mv-inv setup guide in the config (probably at the top) will be good? and maybe share list wiki page at the shares part as well.

nicegamer7 commented:
I'm not sure copying the config would be the best idea... since the plugin would keep working, it's possible one might not notice. I think we could disable the plugin. That way, the server owner would be forced to check their console to see what's wrong.

For the header, I don't really like `# ===[ Multiverse Inventories Config ]===` (just personal preference). But I do like the idea of adding the Wiki link.

benwoo1110 commented:
> I'm not sure copying the config would be the best idea... since the plugin would keep working,

The reason for suggesting copying over is because it can cause a config reset on invalid yaml which have reportedly cause people to lose their current config settings.

nicegamer7 commented:
I think if it fails to load, we should just disable the plugin. Although... now that I think about it, can you disable a plugin after the server has already started?

SlimeDog commented:
> I think if it fails to load, we should just disable the plugin. Although... now that I think about it, can you disable a plugin after the server has already started?

Of course.

benwoo1110 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -66,49 +66,44 @@ public boolean save() {
         if (!doComments) {
             return saved;
         }
```
we can just do:
```java
try {
    config.save(file);
} catch (Exception e) {
    return false;
}
if (!doComments || comments.isEmpty()) {
    return true;
}
```

benwoo1110 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -66,49 +66,44 @@ public boolean save() {
         }
+
         // if there's comments to add and it saved fine, we need to add comments
         if (!comments.isEmpty() && saved) {
```
Then this wont be needed. One less nesting ;)

benwoo1110 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -128,8 +123,8 @@ public boolean save() {
                         }
                         // Find out if the current depth (whitespace * 2) is greater/lesser/equal to the previous depth
                         if (whiteSpace / 2 > depth) {
```
Looks like `whiteSpace / 2` done a vary places, may be a good idea to do it once and store it into a variable for reuse.

nicegamer7 replied:
Yup, good idea.

benwoo1110 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -160,116 +155,85 @@ public boolean save() {
-                line = leadingSpaces + line;
-            } else {
-                line = " ";
+                line = leadingSpaces.toString() + line;
```
Use the StringBuilder of `commentString` here too?

nicegamer7 replied:
Yeah I ended up doing it a bit differently, but same idea.

nicegamer7 replied:
Actually never mind, I get you meant.

nicegamer7 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -13,33 +13,33 @@
+    private final FileConfiguration config;
+    private final boolean doComments;
+    private final HashMap<String, String> comments;
+    private final Pattern newlinePattern = Pattern.compile("\r?\n");
```
Yeah I suppose so.

benwoo1110 commented on src/main/java/com/onarandombox/multiverseinventories/util/CommentedYamlConfiguration.java:
```diff
@@ -160,116 +155,87 @@ public boolean save() {
                     node = false;
                 }
+
                 StringBuilder newLine = new StringBuilder(line);
```
Will it be a better idea to have an empty `StringBuilder` here so we use `newLine.append(...)` instead of `newLine.insert(0, ...)` below? iirc append has better time complexity than insert assuming StringBuilder backend logic is similar to arraylist

nicegamer7 replied:
Just searched it up, you're right. Insert's time complexity is worse than append's.

I'm trying some stuff now to reduce the usage of insert.

benwoo1110 replied:
Yea you should be able to just add the comment lines before the config line, should remove the need for insert for that part.

benwoo1110 commented on src/test/java/com/onarandombox/multiverseinventories/TestCommentedYamlConfiguration.java:
```diff
@@ -0,0 +1,105 @@
+
+public class TestCommentedYamlConfiguration {
+    private static final char LINE_SEPARATOR = '\n';
+    private static final String TEST_CONFIG_NAME = "testconfig.yml";
```
Looks like this file generated during the test isn't ignored by git
